_posts/2021-04-20-custom-errors.md

```diff
@@ -0,0 +1,155 @@
+
+### Errors in Depth
+
+In the above example, `revert Unauthorized();` is equivalent to the following Yul code:
+
+```
+let free_mem_ptr := mload(64)
+mstore(free_mem_ptr, 0x82b4290000000000000000000000000000000000000000000000000000000000)
+revert(free_mem_ptr, 4)
+```
+
+This is the same as the ABI encoding of a function call with the name `Unauthorized()`. Here
+`0x82b42900` is the 'selector' for `Unauthorized()`. In contrast, using a revert string, i.e.,
+`revert("Unauthorized");` leads to the following Yul code:
+
+```
+let free_mem_ptr := mload(64)
+mstore(free_mem_ptr, 0x08c379a000000000000000000000000000000000000000000000000000000000)
+mstore(add(free_mem_ptr, 4), 32)
+mstore(add(free_mem_ptr, 36), 12)
+mstore(add(free_mem_ptr, 68), "Unauthorized")
+revert(free_mem_ptr, 100)
+```
+
+Here `0x08c379a` is the 'selector' of `Error(string)`. In this comparison, one can see that custom
+errors decrease both deploy and runtime gas costs. Note that runtime gas cost is only relevant when
+the revert condition is met.
+
+## Errors with Parameters
+
+It is also possible to have errors that take parameters. For example,
+
+```solidity
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.4;
+
+/// Insufficient balance for transfer. Needed `required` but only
+/// `available` available.
+/// @param available balance available.
+/// @param required requested amount to transfer.
+error InsufficientBalance(uint256 available, uint256 required);
+
+contract TestToken {
+    mapping(address => uint) balance;
+    function transfer(address to, uint256 amount) public {
+        if (amount > balance[msg.sender])
+            // Error call using named parameters. Equivalent to
+            // InsufficientBalance(balance[msg.sender], amount)
+            revert InsufficientBalance({
+                available: balance[msg.sender],
+                required: amount
+            });
+        balance[msg.sender] -= amount;
+        balance[to] += amount;
+    }
+    // ...
+}
+```
+
+The error data would be encoded identically as the ABI encoding for function calls, i.e.,
+``abi.encodeWithSignature("InsufficientBalance(uint256,uint256)", balance[msg.sender], amount)``.
+
+We hope that frameworks will provide support for custom errors. The following is an example on
+decoding error data using [ethers.js](https://docs.ethers.io/v5/):
+
+```javascript
+import { ethers } from "ethers";
+
+// As a workaround, we have a function with the
+// same name and parameters as the error in the abi.
+const abi = [{
+    "inputs": [{
+            "internalType": "uint256",
+            "name": "available",
+            "type": "uint256"
+        },
+        {
+            "internalType": "uint256",
+            "name": "required",
+            "type": "uint256"
+        }
+    ],
+    "name": "InsufficientBalance",
+    "outputs": [],
+    "stateMutability": "nonpayable",
+    "type": "function"
+}];
```

Comment:
One reason  why ethers is far superior to web3.js:

```suggestion
const abi = ["function InsufficientBalance(uint256, uint256)"];
```

Reply:
This would only return

```
[
  BigNumber { _hex: '0x0100', _isBigNumber: true },
  BigNumber { _hex: '0x0100000000', _isBigNumber: true }
]
```

With the full abi-json, it's nice to see the parameters `available` and `required` mapped to values.

Reply:
Does it help if you use `function InsufficientBalance(uint256 available, uint256 required)`?

Reply:
Wow. That actually works. I'll change the abi.